File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) 
    {
        vector<int> left, right;
        int pro = 1;
        for(int i = 0; i<nums.size(); i++)
        {
            left.push_back(pro);
            pro*=nums[i];
        }

        pro = 1;
        for(int i = nums.size()-1; i>=0; i--)
        {
            right.push_back(pro);
            pro*=nums[i];
        }

        for(int i = 0; i<nums.size(); i++)
        {
            left[i] *= right[nums.size() - i - 1];
        }
        return left;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pre = 1
        suff = 1
        res = [1] * n

        for i in range(n):
            res[i] *= pre
            pre *= nums[i]

            res[n - i - 1] *= suff
            suff *= nums[n - i - 1]

        return res